swift/ql/lib/codeql/swift/elements/pattern/Pattern.qll

```diff
@@ -1,4 +1,126 @@
+private import codeql.swift.elements.expr.DeclRefExpr
+private import codeql.swift.elements.stmt.ConditionElement
+private import codeql.swift.elements.stmt.SwitchStmt
+private import codeql.swift.elements.stmt.CaseStmt
+private import codeql.swift.elements.decl.VarDecl
+private import codeql.swift.elements.decl.PatternBindingDecl
+private import codeql.swift.elements.decl.EnumElementDecl
+private import codeql.swift.generated.ParentChild
 
-class Pattern extends Generated::Pattern { }
+class Pattern extends Generated::Pattern {
+  /**
+   * Gets the expression that this top-level pattern is matched against, if any.
+   *
+   * For example, in `switch e { case p: ... }`, the pattern `p`
+   * is immediately matched against the expression `e`.
+   */
+  Expr getImmediateMatchingExpr() {
+    exists(ConditionElement c |
+      c.getPattern() = this and
+      result = c.getInitializer()
+    )
+    or
+    exists(SwitchStmt s |
+      s.getExpr() = result and
+      s.getACase().getALabel().getPattern() = this
+    )
+    or
+    exists(PatternBindingDecl v, int i |
+      v.getPattern(i) = this and
+      result = v.getInit(i)
+    )
+  }
+
+  /**
+   * Gets the expression that this pattern is matched against, if any.
+   * The expression and the pattern need not be top-level children of
+   * a pattern-matching construct, but they must match each other syntactically.
+   *
+   * For example, in `switch .some(e) { case let .some(p): ... }`, the pattern `p`
+   * is matched against the expression `e`.
+   */
+  Expr getMatchingExpr() {
+    result = this.getImmediateMatchingExpr()
+    or
+    exists(Pattern p | p.getMatchingExpr() = result |
+      this = p.(IsPattern).getSubPattern()
+      or
+      this = p.(OptionalSomePattern).getSubPattern()
+      or
+      this = p.(TypedPattern).getSubPattern()
+    )
+    or
+    exists(TuplePattern p, TupleExpr e, int i |
+      p.getMatchingExpr() = e and
+      this = p.getElement(i) and
+      result = e.getElement(i)
+    )
+    or
+    exists(EnumElementPattern p, EnumElementExpr e, int i |
+      p.getMatchingExpr() = e and
+      this = p.getSubPattern(i) and
+      result = e.getArgument(i).getExpr() and
+      p.getElement() = e.getElement()
+    )
+  }
+
+  /** Holds if this pattern is matched against an expression. */
+  final predicate hasMatchingExpr() { exists(this.getMatchingExpr()) }
+
+  /**
+   * Gets the parent pattern of this pattern, if any.
+   */
+  final Pattern getEnclosingPattern() {
+    result = this.getFullyUnresolved().(Pattern).getImmediateEnclosingPattern()
+  }
+
+  /**
+   * Gets the parent pattern of this pattern, but only if
+   * both patterns match against the same object, which is not
+   * the case when destructuring a complex data structure,
+   * like a tuple or enum.
+   */
+  final Pattern getIdentityPreservingEnclosingPattern() {
+    result = this.getFullyUnresolved().(Pattern).getImmediateIdentityPreservingEnclosingPattern()
+  }
```

Comment:
As far as I can tell, this is only used in dataflow. Could we maybe move the predicate to `DataFlowPrivate`? I realize we may need to move `getImmediateIdentityPreservingEnclosingPattern` as well, and that risks introducing a small amount of duplication. But I think the concept of an identity preserving pattern isn't a concept we need to expose users to at the AST layer.

Reply:
Agreed.

Reply:
Removed it altogether, and inlined in the one place it was used elsewhere.